os/drivers/testcase/test_group.c

```diff
@@ -0,0 +1,256 @@
+{
+	int ret;
+	struct child_status_s child_dummy;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+	if (child->flink != NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child->flink = &child_dummy;
+	group_freechild(child);
+	if (child->flink == &child_dummy) {
+		dbg("group_freechild failed.");
+		group_freechild(child);
+		return ERROR;
+	}
+	return OK;
+}
+
+static int test_group_exit_child(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child_pid = kernel_thread("group", SCHED_PRIORITY_DEFAULT, TASK_STACKSIZE, group_exitchild_func, (char *const *)NULL);
+	if (child_pid < 0) {
+		dbg("task_create failed.");
+		return ERROR;
+	}
+
+	child = group_findchild(group, child_pid);
+	if (child == NULL) {
+		dbg("child is null.");
+		return ERROR;
+	}
+
+	sleep(3);
+
+	child_returned = group_exitchild(group);
+	if (child != child_returned) {
+		dbg("group_exitchild failed.");
+		return ERROR;
+	}
+
+	child_returned = group_removechild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_removechild failed.");
+		return ERROR;
+	}
+
+	group_freechild(child);
+	return OK;
+}
+
+static int test_group_removechildren(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child_pid = -1;
+	child->ch_flags = TCB_FLAG_TTYPE_TASK;
+	child->ch_pid = child_pid;
+	child->ch_status = 0;
+	/* Add the entry into the TCB list of children */
+	group_addchild(group, child);
+
+	/* cross check starts */
+	child_returned = group_findchild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_findchild failed.");
+		goto errout_with_addchild;
+	}
+
+	group_removechildren(group);
+	if (group->tg_children != NULL) {
+		dbg("group_removechildren failed.");
+		goto errout_with_addchild;
+	}
+	group_freechild(child);
+	return OK;
+
+errout_with_addchild:
+	group_removechildren(group);
+	group_freechild(child);
+	return ERROR;
+}
+#endif
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int test_group(int cmd, unsigned long arg)
+{
+	int ret = -EINVAL;
+	switch (cmd) {
+#if defined(CONFIG_SCHED_HAVE_PARENT) && defined(CONFIG_SCHED_CHILD_STATUS)
+	case TESTIOC_GROUP_ADD_FINED_REMOVE_TEST:
+		ret = test_group_add_fined_remove(cmd, arg);
+		break;
+	case TESTIOC_GROUP_ALLOC_FREE_TEST:
+		ret = test_group_alloc_free(cmd, arg);
+		break;
+	case TESTIOC_GROUP_EXIT_CHILD_TEST:
+		ret = test_group_exit_child(cmd, arg);
+		break;
+	case TESTIOC_GROUP_REMOVECHILDREN_TEST:
+		ret = test_group_removechildren(cmd, arg);
+		break;
+#endif
+	}
```

Comment:
Don't you need `default`?

Reply:
I think default case is not necessary.
Default case will return -EINVAL value but ret variable has already -EINVAL value.
Is this coding rule violation ?